dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusPTX.java

```diff
@@ -326,32 +317,18 @@ public Object visitFenceAlias(LitmusPTXParser.FenceAliasContext ctx) {
     }
 
     @Override
-    public Object visitBarrierSync(LitmusPTXParser.BarrierSyncContext ctx) {
-        int barID = ctx.barID().id;
-        if (barID < 0 || barID > 15) {
-            throw new ParsingException("Logical barrier resource out of range (0..15): " + barID);
-        }
-        Fence fence = EventFactory.newFence(ctx.getText().toLowerCase());
-        programBuilder.markBeforeBarrier(mainThread, barID);
-        this.barriers.put(mainThread, barID);
+    public Object visitBarrierSyncConstant(LitmusPTXParser.BarrierSyncConstantContext ctx) {
+        IConst constant = (IConst) ctx.constant().accept(this);
+        Fence fence = EventFactory.PTX.newFenceWithId(ctx.getText().toLowerCase(), constant);
+        fence.addFilters(Tag.PTX.BAR);
         return programBuilder.addChild(mainThread, fence);
     }
 
     @Override
-    public Object visitBarrierArrive(LitmusPTXParser.BarrierArriveContext ctx) {
-        int barID = ctx.barID().id;
-        if (barID < 0 || barID > 15) {
-            throw new ParsingException("Logical barrier resource out of range (0..15): " + barID);
-        }
-        Fence fence = EventFactory.newFence(ctx.getText().toLowerCase());
-        programBuilder.markBeforeBarrier(mainThread, barID);
+    public Object visitBarrierSyncRegister(LitmusPTXParser.BarrierSyncRegisterContext ctx) {
+        Register register = programBuilder.getOrCreateRegister(mainThread, ctx.register().getText(), getArchPrecision());
+        Fence fence = EventFactory.PTX.newFenceWithId(ctx.getText().toLowerCase(), register);
+        fence.addFilters(Tag.PTX.BAR);
```

Comment:
Obsolete once you merge the grammar rule

Reply:
Accepted